### ISSUE TITLE
Child tasks don't cancel their group's scope correctly when they raise an exception on asyncio

### ISSUE THREAD

**Issue by gschaffner** [agronholm/anyio#787]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

master (d1aea98), as well as the latest #774 (f9a1e1a)

### Python version

3.12.6, CPython

### What happened?

There are actually two separate bugs here that (I believe) require two separate fixes, but I am putting them both in this issue since they are so closely related (and to avoid spamming the issue tracker).

1.  On asyncio, a child task can finish with an unhandled exception without cancelling its group's scope.

1.  On asyncio, there is a delay between a child task finishing with an unhandled exception and its group's scope getting cancelled. This does not match the behavior on Trio. Per Trio's documentation:

    > If any task inside the nursery finishes with an unhandled exception, then the nursery **immediately** cancels all the tasks inside the nursery. (emphasis added)

### How can we reproduce the bug?

Here are two tests. They all fail on asyncio and pass on trio. The first test is just a weaker version of the second test, but I have separated the two here because I think it is easier to follow that way.

The weaker case:

```python
# Easy mode: `tg.cancel_scope` gets shielded many event loop cycles after the child task
# exits with an exception.
async def test_shield_after_task_failed_weak() -> None:
    async def taskfunc() -> None:
        raise Exception("child task failed")

    with pytest.raises(BaseExceptionGroup) as exc:
        with CancelScope() as outer_scope:
            async with create_task_group() as tg:
                outer_scope.cancel()
                tg.start_soon(taskfunc)
                with CancelScope(shield=True):
                    await wait_all_tasks_blocked()
                    # Wait at least one more scheduling round to ensure that taskfunc's
                    # task_done on asyncio has finished.
                    await sleep(0.1)
                tg.cancel_scope.shield = True
                assert tg.cancel_scope.cancel_called

    assert len(exc.value.exceptions) == 1
    assert str(exc.value.exceptions[0]) == "child task failed"
```

This weaker test is a regression test only for bug (1); it does not test for bug (2).

To fix this case I believe we just need to change the `.cancel` call in `task_done` to be unconditional here:

https://github.com/agronholm/anyio/blob/d1aea98c4f552d04458e49f1484b9c018f2a3ca8/src/anyio/_backends/_asyncio.py#L733-L734

The stronger case:

```python
# Hard mode: tg.cancel_scope gets shielded after the child task exits with an exception
# but before its task_done (on asyncio) runs.
async def test_shield_after_task_failed() -> None:
    taskfunc_exited = anyio.Event()

    async def taskfunc() -> None:
        try:
            raise Exception("child task failed")
        finally:
            taskfunc_exited.set()
            # (Technically we are setting this event slightly before taskfunc exits, but
            # this won't wake any waiters until the event loop's next batch of task
            # steps/callbacks, which will be the first batch to run after taskfunc has
            # exited.)

    with pytest.raises(BaseExceptionGroup) as exc:
        with CancelScope() as outer_scope:
            async with create_task_group() as tg:
                outer_scope.cancel()
                tg.start_soon(taskfunc)
                with CancelScope(shield=True):
                    # Trio documents: "If any task inside the nursery finishes with an
                    # unhandled exception, then the nursery *immediately* cancels all
                    # the tasks inside the nursery" (emphasis added). So when the event
                    # loop drives taskfunc's coro one step forward and it exits with an
                    # exception, tg.cancel_scope must get cancelled *immediately*, i.e.
                    # it must get cancelled before any other tasks/callbacks get to run
                    # a step.
                    await taskfunc_exited.wait()
                tg.cancel_scope.shield = True
                assert tg.cancel_scope.cancel_called

    assert len(exc.value.exceptions) == 1
    assert str(exc.value.exceptions[0]) == "child task failed"
```

This stronger case is a regression test for both bug (1) and bug (2).

(It's a bit hard to read because the test has to be pretty careful to be able to deterministically hit the problematic scheduling order.)

To fix this case I believe we need to move the exception-checking logic out of `task_done` and into a wrapper function:

```python
# in _spawn:
@wraps(func)
async def wrapper():
    try:
        return await func(...)
    finally:
        # task_done's exception-checking logic goes here, i.e. (simplified pseudocode)
        if exception:
            self.cancel_scope.cancel()
```

This way the `.cancel` call will happen **immediately** when the task finishes with an exception, as Trio documents, rather than the `.cancel` being postponed one scheduling batch later.

(Note, this test can also be written equivalently (I believe) using `wait_all_tasks_blocked` in order to hit the problematic scheduling order rather than using an event as I did above. I find the event-based version preferable (easier to understand), but the following is (I think) an equivalent way to implement this test:

```python
async def test_shield_after_task_failed2() -> None:
    async def taskfunc() -> None:
        raise Exception("child task failed")

    with pytest.raises(BaseExceptionGroup) as exc:
        with CancelScope() as outer_scope:
            async with create_task_group() as tg:
                outer_scope.cancel()
                tg.start_soon(taskfunc)
                with CancelScope(shield=True):
                    # Trio documents: "If any task inside the nursery finishes with an
                    # unhandled exception, then the nursery *immediately* cancels all
                    # the tasks inside the nursery" (emphasis added). So when the event
                    # loop drives taskfunc's coro one step forward and it exits with an
                    # exception, tg.cancel_scope must get cancelled *immediately*, i.e.
                    # it must get cancelled before any other tasks/callbacks get to run
                    # a step.
                    await wait_all_tasks_blocked()
                tg.cancel_scope.shield = True
                assert tg.cancel_scope.cancel_called

    assert len(exc.value.exceptions) == 1
    assert str(exc.value.exceptions[0]) == "child task failed"
```

)

**Comment by agronholm** [comment]
I've never seen anyone use cancel scope shielding on a task group. I think even exposing it directly is somewhat questionable. Can you clarify if this is needed for the bug to manifest itself?

**Comment by gschaffner** [comment]
I have moved the following comment from https://github.com/agronholm/anyio/pull/1093#issuecomment-4084397952 to this thread:

> We're hitting this in production:
>
> **Stack:** anyio 4.12.1, CPython 3.12, [LangGraph Platform](https://github.com/langchain-ai/langgraph) (`langgraph-api` 0.7.4)
>
> **LangGraph uses anyio task groups internally for streaming.**
>
> When an outer scope gets cancelled (HTTP disconnect/timeout), child tasks that fail don't cancel the group's scope because `_effectively_cancelled` walks the parent chain and returns `True`.
>
> The host task never learns the child failed, and cleanup hits:
>
> ```
> RuntimeError: Attempted to exit cancel scope in a different task than it was entered in
> ```
>
> The minimal fix from @gschaffner's [review comment on #774](https://github.com/agronholm/anyio/pull/774#discussion_r1755340156), removing the `_effectively_cancelled` guard, resolves it. All 394 existing `test_taskgroups.py` tests pass.
>
> @agronholm No, shielding is not required for this to manifest.
>
> We're hitting it in production with no explicit shielding, just nested task groups where an outer scope gets cancelled (HTTP timeout/disconnect). See PR #1093 for the fix and details.
>
>
> **Langraph Workaround:** We monkey-patch `langgraph_api.stream.consume()` to catch and suppress the `RuntimeError` (the agent has already completed by that point), plus a fallback that queries LangGraph's thread state when the SSE stream ends without a final event.
>
>
> This one-line fix PR in `task_done` removes it at the source for the whole ecosystem without needing that messy workaround.
>
> _Originally posted by @danieldorado-cmyk in https://github.com/agronholm/anyio/issues/1093#issuecomment-4084397952_

**Comment by gschaffner** [comment]
Okay! Thanks for the thoughtful/detailed comment. No apology necessary; it happens.

> Our production crash most likely comes from how `langgraph-api` handles async generators inside task groups, where finalization ends up in a different task context during cancellation. Our workaround just suppresses the symptom.

I'm not sure if it would help you here or not, but maybe take a look at https://trio.readthedocs.io/en/stable/reference-core.html#trio.as_safe_channel.

> I think the PR itself is still a valid fix for the bug described in https://github.com/agronholm/anyio/issues/787, though as you say that does require shielding to observe.
Happy to adjust the PR or close it if you'd rather address https://github.com/agronholm/anyio/issues/787 differently.

https://github.com/agronholm/anyio/pull/1093#issuecomment-4095875109

> Our `RuntimeError` is actually a LangGraph bug 
> https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/pregel/_loop.py#L1407-L1423) 
>
>It creates an `exit_task` via `asyncio.create_task()` that ends up exiting cancel scopes entered by a different task.

Yeah, this looks strange. Perhaps some form of shielding for their `AsyncExitStack.__aexit__` call would help fix this there. (That's just conjecture; I did not look at what's parsing that `CancelledError.args` later in much detail.)
